clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -161,12 +204,69 @@ public void authenticate() throws IOException {
             netInBuffer.payload().get(clientToken, 0, clientToken.length);
             netInBuffer = null; // reset the networkReceive as we read all the data.
             try {
-                byte[] response = saslServer.evaluateResponse(clientToken);
-                if (response != null) {
-                    netOutBuffer = new NetworkSend(node, ByteBuffer.wrap(response));
-                    flushNetOutBufferAndUpdateInterestOps();
+                switch (saslState) {
+                    case INIT:
+                        String clientMechanism = null;
+                        SaslMechanismResponse mechanismResponse = null;
+                        SaslException exception = null;
+                        try {
+                            SaslMechanismRequest mechanismRequest = new SaslMechanismRequest(ByteBuffer.wrap(clientToken));
+                            clientMechanism = mechanismRequest.mechanism();
+                            if (enabledMechanisms.contains(clientMechanism)) {
+                                LOG.debug("Using SASL mechanism '{}' provided by client", clientMechanism);
+                                mechanismResponse = new SaslMechanismResponse((short) 0, Collections.<String>emptyList());
+                            } else {
+                                LOG.debug("SASL mechanism '{}' provided by client is not supported", clientMechanism);
+                                mechanismResponse = new SaslMechanismResponse(SaslMechanismResponse.UNSUPPORTED_SASL_MECHANISM, enabledMechanisms);
+                                exception = new SaslException("Unsupported SASL mechanism " + clientMechanism);
+                            }
+                            clientToken = null;
+                        } catch (Exception e) {
```

Comment:
Could we handle an explicit exception due to the first packet not being a MechanismRequest? I was thinking that we can catch all exceptions from SaslMechanismRequest(ByteBuffer buffer) and convert that to a SchemaException.